### PR TITLE
Ensure Nanny doesn't restart workers that fail to start, and joins subprocess

### LINKED ISSUE #6426: Nanny restarts worker once if it fails to start / doesn't clean up processes on startup failure
If the worker fails during startup, the Nanny will try to restart it once. This is not too important, but it's a bug and it makes testing difficult.

If we slightly modify `test_worker_start_exception` to verify that `"Restarting worker"` isn't in the logs...
https://github.com/dask/distributed/blob/4211e5cddbc59e31e338ddcd4e497a68893e6f14/distributed/tests/test_nanny.py#L485-L490

```python
@gen_cluster(nthreads=[])
async def test_worker_start_exception(s):
    with captured_logger(logger="distributed.nanny", level=logging.DEBUG) as logs:
        # make sure this raises the right Exception:
        with raises_with_cause(RuntimeError, None, StartException, None):
            async with Nanny(s.address, worker_class=BrokenWorker) as n:
                pass

    print("HERE")
    print(logs.getvalue())
    assert "Restarting worker" not in logs.getvalue()
    assert False # fail test so we can see stdout
```

...it passes.

Yet if we run it, the message we didn't want to see actually _is_ in the logs (but not in the captured logs)!
(😬 emojis added for visibility to the relevant spots)
```
(dask-distributed) gabe dev/distributed ‹fix-6363*› » pytest distributed/tests/test_nanny.py::test_worker_start_exception -s                                               1 ↵
============================================================================= test session starts =============================================================================
platform darwin -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /Users/gabe/miniconda3/envs/dask-distributed/bin/python
cachedir: .pytest_cache
rootdir: /Users/gabe/dev/distributed, configfile: setup.cfg
plugins: repeat-0.8.0, xdist-2.3.0, asyncio-0.12.0, timeout-1.4.2, forked-1.3.0, cov-3.0.0, rerunfailures-10.0
timeout: 300.0s
timeout method: thread
timeout func_only: False
collected 1 item                                                                                                                                                              

distributed/tests/test_nanny.py::test_worker_start_exception 2022-05-23 16:11:06,515 - distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
2022-05-23 16:11:06,519 - distributed.scheduler - INFO - State start
2022-05-23 16:11:06,522 - distributed.scheduler - INFO - Clear task state
2022-05-23 16:11:06,523 - distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:61231
2022-05-23 16:11:06,523 - distributed.scheduler - INFO -   dashboard at:           127.0.0.1:61230
2022-05-23 16:11:06,531 - distributed.nanny - INFO -         Start Nanny at: 'tcp://127.0.0.1:61232'
2022-05-23 16:11:07,260 - distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 872, in run
    await worker
  File "/Users/gabe/dev/distributed/distributed/tests/test_nanny.py", line 482, in start
    raise StartException("broken")
test_nanny.StartException: broken
2022-05-23 16:11:07,283 - distributed.nanny - ERROR - Failed while trying to start worker process: broken
2022-05-23 16:11:07,284 - distributed.nanny - ERROR - Failed to connect to process
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 681, in start
    msg = await self._wait_until_connected(uid)
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 801, in _wait_until_connected
    raise msg["exception"]
test_nanny.StartException: broken
2022-05-23 16:11:07,285 - distributed.nanny - ERROR - Failed to start process
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 425, in instantiate
    result = await self.process.start()
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 681, in start
    msg = await self._wait_until_connected(uid)
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 801, in _wait_until_connected
    raise msg["exception"]
test_nanny.StartException: broken
2022-05-23 16:11:07,285 - distributed.nanny - INFO - Closing Nanny at 'tcp://127.0.0.1:61232'.
HERE  # <--- 😬😬😬 OUR LOGGING CAPTURE ENDS. `async with` block has exited
        Start Nanny at: 'tcp://127.0.0.1:61232'
Failed while trying to start worker process: broken
Failed to connect to process
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 681, in start
    msg = await self._wait_until_connected(uid)
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 801, in _wait_until_connected
    raise msg["exception"]
test_nanny.StartException: broken
Failed to start process
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 425, in instantiate
    result = await self.process.start()
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 681, in start
    msg = await self._wait_until_connected(uid)
  File "/Users/gabe/dev/distributed/distributed/nanny.py", line 801, in _wait_until_connected
    raise msg["exception"]
test_nanny.StartException: broken
Closing Nanny at 'tcp://127.0.0.1:61232'.

Dumped cluster state to test_cluster_dump/test_worker_start_exception.yaml
2022-05-23 16:11:07,599 - distributed.nanny - INFO - Worker process 37919 was killed by signal 15
2022-05-23 16:11:07,601 - distributed.nanny - WARNING - Restarting worker  # <--- 😬😬😬 WORKER IS RESTARTING
2022-05-23 16:11:07,602 - distributed.scheduler - INFO - Scheduler closing...
2022-05-23 16:11:07,612 - distributed.scheduler - INFO - Scheduler closing all comms
FAILED

================================================================================== FAILURES ===================================================================================
_________________________________________________________________________ test_worker_start_exception _________________________________________________________________________

s = <Scheduler 'tcp://127.0.0.1:61231', workers: 0, cores: 0, tasks: 0>

    @gen_cluster(nthreads=[])
    async def test_worker_start_exception(s):
        with captured_logger(logger="distributed.nanny", level=logging.DEBUG) as logs:
            # make sure this raises the right Exception:
            with raises_with_cause(RuntimeError, None, StartException, None):
                async with Nanny(s.address, worker_class=BrokenWorker) as n:
                    pass
    
        print("HERE")
        print(logs.getvalue())
        assert "Restarting worker" not in logs.getvalue()
>       assert False # fail test so we can see stdout
E       assert False

distributed/tests/test_nanny.py:511: AssertionError
============================================================================ slowest 20 durations =============================================================================
1.11s call     distributed/tests/test_nanny.py::test_worker_start_exception

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
=========================================================================== short test summary info ===========================================================================
FAILED distributed/tests/test_nanny.py::test_worker_start_exception - assert False
============================================================================== 1 failed in 1.37s ==============================================================================
```

### PR DESCRIPTION
**Pull request by gjoseph92**
Closes #6426

I think this could be split into multiple PRs if desired. I'd recommend reviewing commit by commit.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by fjetter** [comment]
Only remotely related to this change but the name `Nanny._on_exit` confused me quite a bit. It took me way too long to realize that this is in fact a *WorkerProcess* on_exit handler and not a *Nanny* on_exit. I was wondering why we'd even restart the worker if the nanny is exiting...

Doesn't need to happen in this PR but we should consider renaming these methods

File: distributed/tests/test_nanny.py [hunk 1]
```diff
@@ -484,10 +486,18 @@ async def start(self):
 
 @gen_cluster(nthreads=[])
 async def test_worker_start_exception(s):
-    # make sure this raises the right Exception:
-    with raises_with_cause(RuntimeError, None, StartException, None):
-        async with Nanny(s.address, worker_class=BrokenWorker) as n:
-            pass
+    nanny = Nanny(s.address, worker_class=BrokenWorker)
```
Comment (fjetter) [inline distributed/tests/test_nanny.py:489]: 1. We should change the `BrokenWorker.start` to `BrokenWorker.start_unsafe`. `Server.start` is no longer supposed to be overridden by subclasses after https://github.com/dask/distributed/pull/5910
2. Instead of capturing the logs, I believe we should do something like

```python
class BrokenWorker(worker.Worker):
    startup_attempts = 0
    async def start_unsafe(self):
        BrokenWorker.startup_attempts += 1
        raise StartException("broken")

def test_foo():
    BrokenWorker.startup_attempts = 0 # in case it is used elsewhere
    try:
        ...
        assert BrokenWorker.startup_attempts == 1
    finally:
        BrokenWorker.startup_attempts = 0 # in case it is used elsewhere

Comment (gjoseph92) [reply distributed/tests/test_nanny.py:489]: Actually, it's not that easy, because the BrokenWorker is started in a subprocess, so the global variable doesn't carry over. We have to use a multiprocessing queue or shared memory.

File: distributed/tests/test_nanny.py [hunk 2]
```diff
@@ -472,21 +474,39 @@ async def test_nanny_closed_by_keyboard_interrupt(ucx_loop, protocol):
...
+        ):
+            async with nanny:
+                pass
+    assert nanny.status == Status.failed
+    # ^ NOTE: `Nanny.close` sets it to `closed`, then `Server.start._close_on_failure` sets it to `failed`
+    assert nanny.process is None
+    assert startup_attempts.value == 1
+    assert "Restarting worker" not in logs.getvalue()
+    # Avoid excessive spewing. (It's also printed once extra within the subprocess, which is okay.)
+    assert logs.getvalue().count("ValueError: broken") == 1, logs.getvalue()
```
Comment (gjoseph92) [inline distributed/tests/test_nanny.py:509]: @fjetter I didn't remove the logging capture yet (though I could 00e277edcb26152dd0a1cebc4193332d886ccd14), because I liked having this assertion. I noticed that previously, the exception was being logged multiple times from the nanny side, which would make things even noisier and harder to decipher for users.

I did add a multiprocessing Value to count the startup attempts, but I'm not really sure this is necessary or not.

File: distributed/nanny.py [hunk 3]
```diff
@@ -744,32 +751,38 @@ async def kill(self, timeout: float = 2, executor_wait: bool = True):
...
+            return
+        except asyncio.TimeoutError:
+            pass
 
-        if process.is_alive():
-            logger.warning(
-                f"Worker process still alive after {timeout} seconds, killing"
-            )
-            try:
-                await process.terminate()
```
Comment (gjoseph92) [inline distributed/nanny.py:770]: Change: we used to first call `Worker.close` (via `child_stop_q.put({"op": "stop"})`), then send SIGTERM to the process if it didn't stop in time, then just return to the caller as soon as the terminate signal was sent, regardless of whether the process actually stopped.

Now, we still do `Worker.close`, but then send SIGKILL (which unlike SIGTERM, can't be blocked) because the `Worker.close` can already be considered our SIGTERM-like request for graceful close. We wait for the process to actually terminate.

File: distributed/nanny.py [hunk 4]
```diff
@@ -744,32 +751,38 @@ async def kill(self, timeout: float = 2, executor_wait: bool = True):
...
-            except Exception as e:
-                logger.error("Failed to kill worker process: %s", e)
+        logger.warning(
+            f"Worker process still alive after {wait_timeout} seconds, killing"
+        )
+        await process.kill()
+        await process.join(max(0, deadline - time()))
```
Comment (gjoseph92) [inline distributed/nanny.py:785]: I'm a little wary of this deadline on the join. I could imagine the default 2s timeout not being long enough for the process to actually shut down in CI. `kill` didn't used to raise an error if the timeout expired; now it will. I could make it not, but I think it's really the caller's responsibility to decide what to do if the process doesn't shut down in time.

I think we should probably make the default timeout a bit longer.

File: distributed/tests/test_nanny.py (hunk 2 above)
Comment (gjoseph92) [reply distributed/tests/test_nanny.py:499]: I've removed the multiprocessing Value. It seems unnecessary. And in one test, restarting the worker repeatedly segfaulted the Python interpreter. I have no evidence for this, but I feel like maybe some broken state in a shared Value could cause that.

File: distributed/core.py [hunk 5]
```diff
@@ -309,6 +309,7 @@ async def start_unsafe(self):
         await self.rpc.start()
         return self
 
+    @final
```
Comment (gjoseph92) [inline distributed/core.py:312]: driveby: added so nobody in the future can try to override this

File: distributed/nanny.py [hunk 6]
```diff
@@ -405,7 +405,7 @@ async def instantiate(self) -> Status:
             self.process = WorkerProcess(
                 worker_kwargs=worker_kwargs,
                 silence_logs=self.silence_logs,
-                on_exit=self._on_exit_sync,
+                on_exit=self._on_worker_exit_sync,
```
Comment (gjoseph92) [inline distributed/nanny.py:408]: driveby: renamed for clarity

**Comment by hendrikmakait** [comment]
It looks like the added tests work now, but the error of `test_quiet_close_process` (#6582) has changed.

**Comment by hendrikmakait** [comment]
Locally, `test_quiet_close_process` still flakes with

```
sys:1: RuntimeWarning: coroutine 'InProc.write' was never awaited
Task was destroyed but it is pending!
task: <Task pending name='Task-70' coro=<InProc.write() running at /Users/hendrikmakait/projects/dask/distributed/distributed/comm/inproc.py:215> cb=[IOLoop.add_future.<locals>.<lambda>() at /opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed-py3.9/lib/python3.9/site-packages/tornado/ioloop.py:688]>
```

**Comment by hendrikmakait** [comment]
@gjoseph92: I'd say we merge this in and monitor for changes in `test_quiet_close_process`? #6551 already aims at solving the local issue.

**Comment by gjoseph92** [comment]
Thanks for finishing this up @hendrikmakait
